### PR TITLE
[ng] new spinner button (#1488)

### PR DESCRIPTION
**Pull request by jeeyun**
Implementation of the new spinner button. After reviewing with @youdz came to a couple of realization which led to the approach in this PR:

- the new implementation should be non-breaking, so need to continue to support [clrLoading] on button with a boolean value input, while allowing newer approach of taking in different button states
- we want to continue to use clrLoading / LoadingListener to not duplicate this logic in button when many other components already utilize this directive.
- clrLoading is now modified to handle more than two states (loading / not loading)

Signed-off-by: Jeeyun Lim <jeeyun.lim@gmail.com>

### CODE REVIEW

File: src/app/buttons/button-loading.ts [hunk 1]
```diff
@@ -4,35 +4,45 @@
  * The full license information can be found in LICENSE in the root directory of this project.
  */
 import {Component} from "@angular/core";
+import {ClrLoadingState} from "../../clr-angular/utils/loading";
```
Comment (youdz) [inline src/app/buttons/button-loading.ts]: The app should always import from `"../../clr-angular"`, since it acts as a consumer of Clarity and anything deeper is not public.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 2]
```diff
@@ -5,26 +5,84 @@
...
+import {Component, ElementRef, EventEmitter, Output, Renderer2} from "@angular/core";
+import {ClrLoadingState, LoadingListener} from "../../utils/loading";
 
 @Component({
     selector: "button[clrLoading]",
     template: `
-        <span class="spinner spinner-inline" *ngIf="loading"></span>
-        <ng-content></ng-content>
+        <ng-container [ngSwitch]="state">
+            <span *ngSwitchCase="buttonState.LOADING" class="spinner-btn-wrapper">
```
Comment (youdz) [inline src/clr-angular/button/button-loading/loading-button.ts]: Did the switch solve your animation issue?

File: src/clr-angular/image/_icons.clarity.scss [hunk 3]
```diff
@@ -21,6 +21,14 @@ $svg_data: "data:image/svg+xml;charset=utf8";
     @return url("#{$svg_data},#{$warning-icon}");
 }
 
+@function generateCheckIcon($fillColor:$gray-dark) {
```
Comment (youdz) [inline src/clr-angular/image/_icons.clarity.scss]: Wait, why are we not using a `<clr-icon>` here? We're trying _not_ to have any encoded SVGs in the CSS anymore.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 4]
```diff
@@ -5,26 +5,84 @@
...
+
+    updateLoadingState(state: ClrLoadingState): void {
+        this.state = state;
+
+        switch (state) {
+            case ClrLoadingState.DEFAULT:
+                this.renderer.removeAttribute(this.el.nativeElement, "disabled");
+                break;
+            case ClrLoadingState.LOADING:
+                const width = getComputedStyle(this.el.nativeElement).getPropertyValue("width");
```
Comment (youdz) [inline src/clr-angular/button/button-loading/loading-button.ts]: I'll let @gnomeontherun confirm, but I believe we want to wrap this in a check for server-side rendering.

File: src/clr-angular/utils/loading/loading-listener.ts [hunk 5]
```diff
@@ -4,12 +4,10 @@
...
  * This is an abstract class because we need it to still be a valid token for dependency injection after transpiling.
  * This does not mean you should extend it, simply implementing it is fine.
  */
-export abstract class LoadingListener {
-    abstract startLoading(): void;
-
-    abstract doneLoading(): void;
-}
\ No newline at end of file
+export abstract class LoadingListener { abstract updateLoadingState(state: ClrLoadingState): void; }
```
Comment (youdz) [inline src/clr-angular/utils/loading/loading-listener.ts]: I know I'm a PITA, but this "update" method feels like calling it will update the loading state itself, but it doesn't. It's what's called _when_ the loading state has been updated.

I feel like `loadingStateChange(state: ClrLoadingState)` shows better what the method should do: update the view based on a new loading state.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 6]
```diff
@@ -5,26 +5,84 @@
...
+                this.renderer.setStyle(this.el.nativeElement, "width", width);
+                this.renderer.setAttribute(this.el.nativeElement, "disabled", "");
+                break;
+            case ClrLoadingState.SUCCESS:
+                setTimeout(() => {
+                    this.updateLoadingState(ClrLoadingState.DEFAULT);
+                }, 1000);
+                break;
+            case ClrLoadingState.ERROR:
+                this.updateLoadingState(ClrLoadingState.DEFAULT);
```
Comment (youdz) [inline src/clr-angular/button/button-loading/loading-button.ts]: You're not updating the actual Loading directive here. Maybe it doesn't matter, but it feels weird. Does that mean we need a second method on `LoadingListener` to allow for communication both ways?

File: src/clr-angular/utils/loading/loading.module.ts [hunk 7]
```diff
@@ -6,13 +6,15 @@
...
 
-import {ClrLoading} from "./loading";
+import {ClrLoading, ClrLoadingState} from "./loading";
 
 export const CLR_LOADING_DIRECTIVES: Type<any>[] = [ClrLoading];
 
 @NgModule({imports: [CommonModule], declarations: [CLR_LOADING_DIRECTIVES], exports: [CLR_LOADING_DIRECTIVES]})
 export class ClrLoadingModule {}
 
+export const loadingState = ClrLoadingState;
```
Comment (youdz) [inline src/clr-angular/utils/loading/loading.module.ts]: Not sure what this renamed export is for. I might have missed it somewhere. But it looks like it can be safely removed.

File: src/clr-angular/utils/loading/loading.ts [hunk 8]
```diff
@@ -7,32 +7,46 @@ import {Directive, Input, OnDestroy, Optional} from "@angular/core";
...
             return;
         }
-        this._loading = value;
-        if (this.listener) {
+
+        if (value === null) {
+            value = false;
+        }
+
+        if (typeof value === "boolean") {
```
Comment (youdz) [inline src/clr-angular/utils/loading/loading.ts]: This if should go _before_ the first one, so that setting `true` when the state is already `LOADING` is ignored too.

File: src/app/buttons/button-loading.ts (hunk 1 above)
Comment (jeeyun) [reply src/app/buttons/button-loading.ts]: 👍

File: src/clr-angular/button/button-loading/loading-button.ts (hunk 2 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: Nope, unfortunately the same results as using `*ngIf`.

File: src/clr-angular/image/_icons.clarity.scss (hunk 3 above)
Comment (jeeyun) [reply src/clr-angular/image/_icons.clarity.scss]: I'm making use of existing `spinner`, `spinner-inline` classes and it makes it a breaking change to get rid of using SVG completely in spinner (or alternatively, gets pretty messy trying to make it an exception for the checkbox case here only). Would it be ok to create a separate issue for getting rid of encoded SVGs in spinner-related components all together in the future?

File: src/clr-angular/utils/loading/loading-listener.ts (hunk 5 above)
Comment (jeeyun) [reply src/clr-angular/utils/loading/loading-listener.ts]: No problem. Will change the name.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 9]
```diff
@@ -59,30 +59,38 @@ export class ClrLoadingButton implements LoadingListener {
 
     constructor(private el: ElementRef, private renderer: Renderer2) {}
 
-    updateLoadingState(state: ClrLoadingState): void {
+    loadingStateChange(state: ClrLoadingState): void {
         this.state = state;
 
         switch (state) {
             case ClrLoadingState.DEFAULT:
+                this.renderer.removeStyle(this.el.nativeElement, "width");
```
Comment (youdz) [inline src/clr-angular/button/button-loading/loading-button.ts]: Why did you remove this? Does it mean the button never reverts to pure dynamic width?

File: src/clr-angular/image/_icons.clarity.scss (hunk 3 above)
Comment (gnomeontherun) [reply src/clr-angular/image/_icons.clarity.scss]: If there is a deprecation we want to make, let's get a ticket in to have that done for 0.12.

File: src/clr-angular/button/button-loading/loading-button.module.ts [hunk 10]
```diff
@@ -7,15 +7,16 @@
...
-import {ClrLoadingModule} from "../../utils/loading/loading.module";
+import {ClrIconModule} from "../../icon";
+
 import {ClrLoadingButton} from "./loading-button";
 
 export const CLR_LOADING_BUTTON_DIRECTIVES: Type<any>[] = [ClrLoadingButton];
 
 @NgModule({
-    imports: [CommonModule, ClrLoadingModule],
+    imports: [CommonModule, ClrIconModule],
```
Comment (gnomeontherun) [inline src/clr-angular/button/button-loading/loading-button.module.ts]: Why do you need `ClrIconModule`?

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 11]
```diff
@@ -5,26 +5,92 @@
...
+            <span *ngSwitchCase="buttonState.DEFAULT" @defaultButton>
+                <ng-content></ng-content>
+            </span>
+        </ng-container>
     `,
-    providers: [{provide: LoadingListener, useExisting: ClrLoadingButton}]
+    providers: [{provide: LoadingListener, useExisting: ClrLoadingButton}],
+    animations: [
+        trigger("defaultButton",
+                [
```
Comment (gnomeontherun) [inline src/clr-angular/button/button-loading/loading-button.ts]: Is this formatting just what clang does automatically? I think its hard to read, if we can do anything about it.

File: src/clr-angular/button/button-loading/loading-button.module.ts (hunk 10 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.module.ts]: Oh this was when I was testing out using clr-icon for the spinner graphics. Will remove.

File: src/clr-angular/button/button-loading/loading-button.ts (hunk 11 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: Some of it is clang. Let me see if I can improve by at least separating the comment into a new line.

File: src/app/buttons/button-loading.html [hunk 12]
```diff
@@ -6,23 +6,23 @@
   -->
 
 <h4>Loading Buttons</h4>
-<button [clrLoading]="validateLoading" class="btn btn-info-outline" (click)="validateDemo()">Validate</button>
-<button [clrLoading]="submitLoading" type="submit" class="btn btn-success-outline" (click)="submitDemo()">Submit</button>
+<button [clrLoading]="validateState" class="btn btn-primary" (click)="validateDemo()">Validate</button>
+<button [clrLoading]="submitState" type="submit" class="btn btn-success-outline" (click)="submitDemo()">Submit</button>
 
 <pre>
```
Comment (adityarb88) [inline src/app/buttons/button-loading.html]: Can we remove the code snippets from the demos? We have decided to do that across all demos now. We will also get rid of prismjs from clarity & our dev app.

File: src/app/buttons/button-loading.ts [hunk 13]
```diff
@@ -4,35 +4,45 @@
...
+            this.validateState = ClrLoadingState.SUCCESS;
+        }, 1500);
     }
 
     submitDemo() {
-        this.submitLoading = true;
-        setTimeout(() => this.submitLoading = false, 1500);
+        this.submitState = ClrLoadingState.LOADING;
+        setTimeout(() => {
+            this.submitState = ClrLoadingState.DEFAULT;
```
Comment (adityarb88) [inline src/app/buttons/button-loading.ts]: What is the difference between changing from `LOADING` to `SUCCESS` like you did in the `validateDemo()` method vs changing from `LOADING` to `DEFAULT`?

File: src/app/buttons/button-loading.ts [hunk 14]
```diff
@@ -4,35 +4,45 @@
...
+        }, 1500);
     }
 
     validateSmDemo() {
-        this.validateSmLoading = true;
-        setTimeout(() => this.validateSmLoading = false, 1500);
+        this.validateSmState = true;
+        setTimeout(() => {
+            this.validateSmState = false;
```
Comment (adityarb88) [inline src/app/buttons/button-loading.ts]: Why is this different than the other methods? I am assuming that we are testing that the old API still works.

File: src/clr-angular/button/button-loading/loading-button.module.ts [hunk 15]
```diff
@@ -7,16 +7,12 @@
...
 
 export const CLR_LOADING_BUTTON_DIRECTIVES: Type<any>[] = [ClrLoadingButton];
 
-@NgModule({
-    imports: [CommonModule, ClrLoadingModule],
-    declarations: [CLR_LOADING_BUTTON_DIRECTIVES],
-    exports: [CLR_LOADING_BUTTON_DIRECTIVES, ClrLoadingModule]
-})
+@NgModule(
+    {imports: [CommonModule], declarations: [CLR_LOADING_BUTTON_DIRECTIVES], exports: [CLR_LOADING_BUTTON_DIRECTIVES]})
```
Comment (adityarb88) [inline src/clr-angular/button/button-loading/loading-button.module.ts]: Is `ClrLoadingModule` no longer needed in the imports?

File: src/clr-angular/button/button-group/button.spec.ts [hunk 16]
```diff
@@ -151,22 +151,21 @@ export default function(): void {
...
+                componentInstance.button1.loadingStateChange(ClrLoadingState.LOADING);
 
                 expect(componentInstance.button1.loading).toBe(true);
             });
 
             it("sets loading to false on done loading", () => {
                 expect(componentInstance.button1.loading).toBeUndefined();
 
-                componentInstance.button1.doneLoading();
+                componentInstance.button1.loadingStateChange(ClrLoadingState.DEFAULT);
```
Comment (adityarb88) [inline src/clr-angular/button/button-group/button.spec.ts]: Are we missing a test for `ClrLoadingState.SUCCESS`?

File: src/app/buttons/button-loading.ts (hunk 13 above)
Comment (adityarb88) [reply src/app/buttons/button-loading.ts]: Nvm. I should have read `loading-button.ts` first :D

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 17]
```diff
@@ -5,26 +5,91 @@
...
+import {ClrLoadingState, LoadingListener} from "../../utils/loading";
 
 @Component({
     selector: "button[clrLoading]",
     template: `
-        <span class="spinner spinner-inline" *ngIf="loading"></span>
-        <ng-content></ng-content>
+        <ng-container [ngSwitch]="state">
+            <span *ngSwitchCase="buttonState.LOADING">
+                <span @spinner class="spinner spinner-inline spinner-loading"></span>
```
Comment (adityarb88) [inline src/clr-angular/button/button-loading/loading-button.ts]: What does the `.spinner-loading` class do and where is it declared? I dont see it in the CSS.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 18]
```diff
@@ -5,26 +5,91 @@
...
+            </span>
+            <span *ngSwitchCase="buttonState.SUCCESS">
+                <span @validated class="spinner spinner-inline spinner-loading spinner-check"></span>
```
Comment (adityarb88) [inline src/clr-angular/button/button-loading/loading-button.ts]: `spinner-check` seems a bit weird to me. Couldn't we have just extended the `.spinner` styles and created a different class or even better, used the `clr-icon` here? We would just have to align that icon correctly in the button, no? Is the `.spinner` class adding any styles to the `checkmark` icon?

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 19]
```diff
@@ -5,26 +5,91 @@
...
+                    transition(":enter", [style({opacity: 0}), animate("200ms 100ms ease-in", style({opacity: 1}))]),
+                    // TODO: see if we can get leave animation to work before spinner's enter animation
+                    transition(":leave", [style({opacity: 0})])
+                ]),
+        trigger("spinner",
+                [
+                    transition(":enter", [style({opacity: 0}), animate("200ms 100ms ease-in", style({opacity: 1}))]),
+                    transition(":leave", [style({opacity: 1}), animate("100ms ease-out", style({opacity: 0}))])
+                ]),
+        trigger("validated",
```
Comment (adityarb88) [inline src/clr-angular/button/button-loading/loading-button.ts]: should we rename this state to `done` or `completed` instead of `validated`?

File: src/app/buttons/button-loading.ts (hunk 14 above)
Comment (jeeyun) [reply src/app/buttons/button-loading.ts]: Yes that's correct. To make sure it's a non-breaking change for old API.

File: src/clr-angular/button/button-loading/loading-button.module.ts (hunk 15 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.module.ts]: Everything in `ClrLoadingModule` is deprecated. I believe it is exported thru `utils` now and you can simply import from `@clr/angular`.

File: src/clr-angular/button/button-loading/loading-button.ts (hunk 17 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: Hmm.. maybe a class I was using while implementing and no longer needed. Will doublecheck and remove if that's the case.

File: src/clr-angular/button/button-loading/loading-button.ts [hunk 20]
```diff
@@ -5,26 +5,91 @@
...
+                [
+                    transition(":enter",
+                               [animate("300ms", keyframes([
+                                            style({transform: "scale(0,0)"}), style({opacity: 1}),
+                                            style({transform: "scale(1.2,1.2)"}), style({transform: "scale(.9,.9)"}),
+                                            style({transform: "scale(1,1)"})
+                                        ]))]),
+                    transition(":leave", [style({opacity: 1}), animate("100ms ease-out", style({opacity: 0}))])
```
[the remark this answers is not in the document]
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: 😄

File: src/clr-angular/button/button-loading/loading-button.ts (hunk 18 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: `spinner-check` leans on the styles of `.spinner` and `.spinner-inline` and just overrides the background and animation. Using `clr-icon` would require breaking away from using any spinner classes since the `animation` and `background`are used there. Changing `spinner` class to use `clr-icon` is a breaking change for static spinner markup, so that was not an option. I created a separate ticket to look into getting rid of embedded SVGs from our components in the future: https://github.com/vmware/clarity/issues/2241. Let me know if my explanation is confusing. Happy to discuss on a call further.

File: src/clr-angular/button/button-loading/loading-button.ts (hunk 19 above)
Comment (jeeyun) [reply src/clr-angular/button/button-loading/loading-button.ts]: Hmm.. I'm torn about that. Depending on the use case going from SPINNING to DEFAULT is considered "done" in some cases even though we skip the `validated` state.

File: src/app/buttons/button-loading.html (hunk 12 above)
Comment (jeeyun) [reply src/app/buttons/button-loading.html]: Yup will remove!

File: src/clr-angular/button/button-group/button.spec.ts (hunk 16 above)
Comment (jeeyun) [reply src/clr-angular/button/button-group/button.spec.ts]: More extensive testing is done in `loading-button.spec.ts` including test for `ClrLoadingState.SUCCESS` :)

File: src/clr-angular/image/_icons.clarity.scss (hunk 3 above)
Comment (jeeyun) [reply src/clr-angular/image/_icons.clarity.scss]: issue created, though not sure this is amust for 0.12: https://github.com/vmware/clarity/issues/2241
